### PR TITLE
adds check so we only queue one copy of the async wakeup request

### LINKED ISSUE #1951: multiple `uv_async_t` notifications can corrupt `loop->pending_requests_tail` list
When `uv_async_send` is called it [calls `POST_COMPLETION_FOR_REQ(loop, &loop->async_req)`](https://github.com/libuv/libuv/blob/33b6db3909dca45fa543ab6405b934f1333fd213/src/win/async.c#L63) If there are two different `uv_async_t` handles that both get triggered, this will happen for both of them and you end up with two different completions that share the same request. When the event loop wakes up and `GetQueuedCompletionStatusEx` is called, it [calls `uv_insert_pending_req`](https://github.com/libuv/libuv/blob/7284adfa7a833264fc67d68d39610da27ecbab7c/src/win/core.c#L396-L412) twice with the same request pointer, which corrupts the pending requests list `loop->pending_requests_tail`. It [looks like](https://github.com/libuv/libuv/blob/87df1448a48fb64c2b9ebe37e3344f7e5b81dd88/src/win/req-inl.h#L93-L102) the possibility of this type of corruption is known.

I came across this while trying to figure out why my callbacks weren't getting called reliably in an application that has pretty high-frequency async notifications from a separate thread. I don't have a separate MWE at the moment - LMK if the issue doesn't seem clear and maybe I can put something together.

This is on windows, using the Julia fork of libuv.

### PR DESCRIPTION
**Pull request by ssfrr**
fixes #1951

### CODE REVIEW

**Comment by ssfrr** [comment]
sorry about that - fixed the spaces.

**Comment by bzoz** [comment]
Could you please add a test for this?

**Comment by bzoz** [comment]
I'm -1. The issue was caused by a change to asyncs internals in https://github.com/libuv/libuv/pull/980. It looks like this PR would just cover that bug and not fix it.

**Comment by ssfrr** [comment]
It seems like there are three main ways to address this issue:
1. add a check during `uv_async_send` to make sure we only post one completion even if there are multiple calls in a single event loop run
2. check while we're handling completions to make sure we only enqueue one request (this PR)
3. revert parts of #980 to go back to having each `uv_async_t` have its own `uv_req_t` that gets posted

I went with (2) because it seemed like the simplest change. (3) seems like a cleaner way to handle things, and would remove the need for some of the queue juggling. I don't really understand enough about that PR to know why it made that change, but this part of the commit message is helpful:

> This change is intended to make uv_async_send more resilient. It has to be thread-safe, and that means that the handle could just have been closed when uv_async_send was called. This case was previously not handles (there is an inherent race condition). The new model is inspired by the one used on the Unix side, which uses a single fd (or overlapped in this case) to wakeup the loop and then process all pending async handles. This makes handling those edge cases a lot simpler: when the handle is closed it's removed from the handle queue, and then it's not processed at all.

I admit I don't fully understand the consequences of `uv_async_send` being called on a handle that's already been closed. Obviously it can't have been `free`d, but that seems like a reasonable restriction on the application that handle memory lives as long as the threads that might access it.

**Comment by vtjnash** [comment]
This lgtm. @ssfrr can you add a test?

@bzoz I think this fixes the bug, and doesn't just make it less likely / obscure it. Comparing against the unix implementation, this appears to continue the changes in #980 to make the windows implementation more similar by ensuring we just discard repeated events.

**Comment by ssfrr** [comment]
I have [MWE C code](https://gist.github.com/ssfrr/ca4a36428c4985cfe329bfad2d9b2a3c) that reproduces the issue, but I had trouble getting the test suite running in windows when I was working on this before. IIRC I couldn't get the libuv build system working under mingw, and to build my MWE and test the PR I actually used Julia's Makefile.

Does anyone have an MSVS setup handy that could maybe adapt the MWE?

**Comment by bnoordhuis** [comment]
@ssfrr The cmake build should work.

**Comment by vtjnash** [comment]
Not stale, but does need a test. It's a critical bugfix, so it's in active use over at JuliaLang.

**Comment by cjihrig** [comment]
Worth noting that this currently has an explicit -1 from @bzoz. I think we would want to resolve that before this could land.

**Comment by vtjnash** [comment]
I know, that's what's been delaying this (or I'd have assumed I should just go ahead and try to land this bugfix). I'm not sure exactly his thinking from the short -1, but that PR he referenced moved the queue into the loop without moving the corresponding atomic status bit. This PR would fix by making it safe to dequeue the event multiple times (like unix). An alternative fix would be to have a copy of `handle->async_sent` on the loop itself (like how #980 moved handle->async_req to the loop) near https://github.com/libuv/libuv/pull/980/files#diff-06f73fca0085a7765805037cad3d1ca9R61

**Comment by saghul** [comment]
@vtjnash This on master though. Is Julia currently using master?

**Comment by vtjnash** [comment]
Yeah, I think I'm the de-facto maintainer of master as a result. I'd say we've spent relatively little time on v1.x (July 2015 to Nov 2017 according to our commit logs) as a combination of a somewhat deferred and bumpy upgrade from v0.11 and a rapid need for the v2 features to cut down our patch list.

**Comment by saghul** [comment]
So, I took a quick look and wondered what saghul from 3 years ago was up to :-) I don't think a revert helps, because that change fixed other issues too.

I think the patch here is ok TBH. If we used a queue for requests we could do a QUEUE_REMOVE and then a QUEUE_INSERT_TAIL so no requuest is queued twice, but since that's is not possible (without further refactor, that is) I find the solution acceptable.

The comment needs to use /* */ and all that, but the approach LGTM.

**Comment by vtjnash** [comment]
Since this is a fairly critical regression for folks using libuv master, I'd like to land this fairly soon. @saghul did you want to make your review official?

CI: https://ci.nodejs.org/job/libuv-test-commit/1698/

**Comment by vtjnash** [comment]
Oops, realized that this didn't have a test added yet. I've done that now (adapted from the MWE above), but someone else should review now to make sure that that looks good.

CI: https://ci.nodejs.org/job/libuv-test-commit/1701/

**Comment by vtjnash** [comment]
Review bump request (of the test style specifically) 🙂

File: test/test-async-multi.c [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+
+void prepare_cb(uv_prepare_t* handle) {
+  uv_pipe_connect(&connect_req, &pipe_dummy, "nonexistent_file_path", connect_cb);
+  uv_close((uv_handle_t*) handle, close_cb);
+}
+
+
+TEST_IMPL(async_multi) {
+  uv_loop_t *loop;
```
Comment (santigimeno) [inline test/test-async-multi.c]: style: star to lean left

File: test/test-async-multi.c [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  uv_async_t async1, async2;
```
Comment (santigimeno) [inline test/test-async-multi.c]: Shouldn't these be global?

File: test/test-async-multi.c [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+  loop = uv_default_loop();
+
+  ASSERT(0 == uv_async_init(loop, &async1, async_cb));
+  ASSERT(0 == uv_async_init(loop, &async2, async_cb));
+
+  /* Create a pending notification */
+  ASSERT(0 == uv_pipe_init(loop, &pipe_dummy, 0));
+  ASSERT(0 == uv_prepare_init(uv_default_loop(), &prepare));
```
Comment (santigimeno) [inline test/test-async-multi.c]: nit: use `loop` directly

File: test/test-async-multi.c (hunk 2 above)
Comment (bnoordhuis) [reply test/test-async-multi.c]: They don't have to be, this works. On the topic of style though:
```suggestion
  uv_async_t async1;
  uv_async_t async2;
```

File: src/win/core.c [hunk 4]
```diff
@@ -318,6 +319,16 @@ static void uv__poll(uv_loop_t* loop, int timeout) {
          */
         if (overlappeds[i].lpOverlapped) {
           req = container_of(overlappeds[i].lpOverlapped, uv_req_t, u.io.overlapped);
+          /* If multiple async handles were triggered we might end up with
+           * multiple UV_WAKEUP requests (IOCP completion events). They all
+           * share the same req however, so we need to be careful to only make
```
Comment (bnoordhuis) [inline src/win/core.c]: I don't understand the "they all share the same req" part. The object that is posted with `PostQueuedCompletionStatus()` is `&handle->async_req.u.io.overlapped`, isn't it?

That's unique for each handle and should (barring bugs) only be posted once until it's processed by `uv_process_reqs()`.

There might be concurrency bugs in our handling of `handle->async_sent` though. 37042a5b fixed one such bug in the unix implementation.

Comment (saghul) [reply src/win/core.c]: @bnoordhuis There is one single async_req per loop: https://github.com/libuv/libuv/blob/7c845fb25f13a03a5264ea28e0a642fc7dbf7c98/src/win/async.c#L63

Comment (bnoordhuis) [reply src/win/core.c]: Oh... it's per-handle in the v1.x branch but per-loop in master. Okay, objection withdrawn.

**Comment by saghul** [comment]
> Since this is a fairly critical regression for folks using libuv master, I'd like to land this fairly soon (I've rebased and addressed review comments). @saghul did you want to make your review official?

Sorry I missed this one. Just did!

**Comment by vtjnash** [comment]
re-CI: https://ci.nodejs.org/job/libuv-test-commit/1734/ (before merge, after fixups, and because I just updated master)

**Pull request merged**